lib/thread/mutex+osx.myr

```diff
@@ -0,0 +1,75 @@
+use std
+use sys
+
+use "atomic"
+
+pkg thread =
+	type mutex = struct
+		_state	: uint64
+	;;	
+
+	const mkmtx	: (-> mutex)
+	const mtxlock	: (mtx : mutex# -> void)
+	const mtxtrylock	: (mtx : mutex# -> bool)
+	const mtxunlock	: (mtx : mutex# -> void)
+
+	pkglocal const Unlocked = 0
+	pkglocal const Locked = 1
+	pkglocal const Contended = 2
+;;
+
+var nspin = 10	/* FIXME: pick a sane number, based on CPU count */
+
+const mkmtx = {
+	-> [._state = Unlocked]
+}
+
+const mtxlock = {mtx
+	var c
+
+	/* 
+	Uncontended case: we get an unlocked mutex, and we lock it.
+	*/
+        c = Locked
+	for var i = 0; i < nspin; i++
+		c = xcas(&mtx._state, Unlocked, Locked) 
+		if c == Unlocked
+			-> void
+		;;
+	;;
+
+	/*
+	Contended case: we set the lock state to Contended. This indicates that there
+	the lock is locked, and we potentially have threads waiting on it, which means
+	that we will need to wake them up.
+	*/
+	if c == Locked
+		c = xchg(&mtx._state, Contended)
+	;;
+
+	while c != Unlocked
+		sys.ulock_wait(sys.Ulockcompareandwait | sys.Ulocknoerrno, &mtx._state, Contended, 0)
```

Comment:
Just wondering what the possible errors are, and what Ulocknoerrno does.

Reply:
None of this is documented by Apple, but at least the kernel is open source. The relevant file is this one: https://github.com/apple/darwin-xnu/blob/0a798f6738bc1db01281fc08ae024145e84df927/bsd/kern/sys_ulock.c

ULF_NO_ERRNO seems to return the error code to userspace instead of setting errno. The possible error codes are EINVAL and EDOM, which should never happen; ESRCH, ENOENT, and EOWNERDEAD, which we shouldn't have to deal with(?); EALREADY, which looks related to enforcing using mutexes, etc. correctly, which we don't currently do; ENOMEM and EINTR, which we handle by just waiting again although maybe we should die in the case of ENOMEM; and ETIMEDOUT, which we will currently never see but will be useful in the future.

Reply:
Sounds good.

Reply:
(Also, let's put in a comment pointing to the code that we used to learn about how this works)